Rework S3 object version ID and marker check

Neon-White opened the pull request:
### Explain the changes
1. Move the version ID check to happen before `read_object_md` is called, which takes care of both namespacestore and backingstore queries - while also potentially eliminating the case in which the store would enter a Rejected state
2. Remove the max-line linter error in `test_s3_ops`
3. Check for empty strings in all API calls containing a version ID
4. Check for empty strings in all API calls containing a version ID marker

### Issues: Fixed #xxx / Gap #xxx
1. [2265562](https://bugzilla.redhat.com/show_bug.cgi?id=2265562)

### Testing Instructions:
1. Upload an object to a bucket
5. Use the AWS CLI (`s3api` with `--version-id ""`) to try and access an existing object's version ID by providing an empty string
6. Verify that NooBaa throws the appropriate error (`An error occurred (InvalidArgument) when calling the GetObject operation: Version id cannot be the empty string`)

- [x] Tests added

Neon-White committed:
Move S3 `get_object` version ID check

Signed-off-by: Ben <belimele@redhat.com>

guymguym commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -35,6 +35,11 @@ async function get_object(req, res) {
         md_params.part_number = part_number;
     }
+
+    if (req.query.versionId?.length === 0) {
```
The same check should also be in head object.
Best add it as a common function in s3_utils to avoid copy-pasta.
Also move it a bit up so it's with the other request query parsing cases.
```js
const version_id = s3_utils.parse_version_id(req.query.versionId);
```

shirady replied:
@guymguym why just head object and get object?
The function `s3_utils.parse_version_id(req);` is general and can be everywhere there is a parameter of `version_id`, for example: DeleteObject, DeleteObjectTagging, GetObjectAcl, PutObjectRetention, PutObjectTagging, RestoreObject (advance case ListObjectVersions - VersionIdMarker), etc.

We can search all `req.query.versionId` in the `src/endpoint/s3/ops`.

guymguym replied:
But I don't think VersionIdMarker in listObjectVersions is the same.

shirady replied:
> But I don't think VersionIdMarker in listObjectVersions is the same.

@Neon-White I think It is something that we would need to check (from [AWS SDK V2 docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectVersions-property)):

> VersionIdMarker — (String)
> Specifies the object version you want to start listing from.

guymguym replied:
NP, I think we can go ahead and replace all refs to req.query.versionId in s3 ops.

But for versionId in xml fields or as marker - I would first check how aws handles empty string there - I don't want to assume I know the answer.

guymguym replied:
@Neon-White 

I tested on aws bucket -

1. delete-objects - does not return InvalidArgument at all with VersionId="", instead it just returns an array of errors with NoSuchVersion per requested version it cannot find - so we should undo the call to parse_version_id() in that op, and let our code continue with that empty string which will return the array of errors as it usually does -

```
❯ aws s3api delete-objects --bucket my-aws-bucket --delete '{"Objects":[ {"Key":"watman.png","VersionId":""}, {"Key":"README.md","VersionId":""} ]}'
{
    "Errors": [
        {
            "Key": "watman.png",
            "VersionId": "",
            "Code": "NoSuchVersion",
            "Message": "The specified version does not exist."
        },
        {
            "Key": "README.md",
            "VersionId": "",
            "Code": "NoSuchVersion",
            "Message": "The specified version does not exist."
        }
    ]
}
```

2. list-object-versions - VersionIdMarker is an input argument to this op and it does contain a versionId so it's not so different than other ops where versionId when passed as argument - I tested it with VersionIdMarker="" and it does reply with InvalidArgument with a similar yet more specific error message "A version-id marker cannot be empty". So this should actually be checked just like any other version id argument  -

```
❯ aws s3api list-object-versions --bucket my-aws-bucket --key-marker "watman.png" --version-id-marker ""

An error occurred (InvalidArgument) when calling the ListObjectVersions operation: A version-id marker cannot be empty.
```

Neon-White committed:
- Move the obj version ID check to S3 utils
- Move the `get_object` check closer to the parsing checks
- Add the version ID check to `head_object`

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
Fix check logic

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
Add unittest

Signed-off-by: Ben <belimele@redhat.com>

guymguym commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -15,6 +15,9 @@ async function get_object(req, res) {
     const agent_header = req.headers['user-agent'];
     const noobaa_trigger_agent = agent_header && agent_header.includes('exec-env/NOOBAA_FUNCTION');
     const encryption = s3_utils.parse_encryption(req);
+    if (s3_utils.version_id_empty_string_check(req.query.versionId)) {
+        throw new S3Error(S3Error.InvalidArgumentEmptyVersionId);
+    }
```
the throw should be inside that function ...

Neon-White committed:
Fix logic to actually make sense

Signed-off-by: Ben <belimele@redhat.com>

shirady commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -35,6 +36,7 @@ async function get_object(req, res) {
     if (part_number) {
         md_params.part_number = part_number;
     }
+
```
if you didn't plan to add the newlines, please remove them.

shirady commented on src/endpoint/s3/s3_utils.js:
```diff
@@ -622,6 +622,16 @@ function encode_uri_unless_already_encoded(uri = '') {
+ * @param {string} version_id
+ */
+function version_id_empty_string_check(version_id) {
+    if (version_id?.length === 0) {
```
Why did you prefer the length check over explicitly checking if the version ID is an empty string?
```suggestion
    if (version_id === '') {
```

guymguym replied:
oh I agree here too, the explicit empty string check is better than fishing for the length.

guymguym replied:
checking for explicit empty string is better and more readable imho, and I'm even ready to fight with google about it.

guymguym commented on src/endpoint/s3/s3_utils.js:
```diff
@@ -622,6 +622,16 @@ function encode_uri_unless_already_encoded(uri = '') {
+ * version_id_empty_string_check throws an error if version_id is an empty string
+ * @param {string} version_id
+ */
+function version_id_empty_string_check(version_id) {
+    if (version_id?.length === 0) {
+        throw new S3Error(S3Error.InvalidArgumentEmptyVersionId);
+    }
+}
```
I think this function is "leaking" its internal behavior and it is unneeded. As I suggested before, I personally would actually make this a "parse" function so that the caller passes the request and gets back the value of the version it from it after validating it. Similar to s3_utils parse_encryption/storage_class/part_number/etc. like this:
```suggestion
function parse_version_id(req) {
    const version_id = req.query.versionId;
    if (version_id?.length === 0) {
        throw new S3Error(S3Error.InvalidArgumentEmptyVersionId);
    }
    return version_id;
}
```

guymguym replied:
I guess I changed my suggestion :)

Neon-White committed:
- Change `parse_version_id` logic

Signed-off-by: Ben <belimele@redhat.com>

guymguym commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -53,15 +55,13 @@ async function get_object(req, res) {
         bucket: req.params.bucket,
         key: req.params.key,
-        version_id: req.query.versionId,
+        version_id: version_id,
```
short property notation FTW -
```suggestion
        version_id,
```

guymguym commented on src/endpoint/s3/ops/s3_head_object.js:
```diff
@@ -11,6 +11,7 @@ const S3Error = require('../s3_errors').S3Error;
  */
 async function head_object(req, res) {
     const encryption = s3_utils.parse_encryption(req);
+    s3_utils.parse_version_id(req.query.versionId);
```
also grab the returned value here and replace refs to req.query.versionId in this function
```suggestion
    const version_id = s3_utils.parse_version_id(req.query.versionId);
```

Neon-White committed:
Utilize `parse_version_id`.... everywhere.

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
Please DeepScan

Signed-off-by: Ben <belimele@redhat.com>

shirady commented on src/endpoint/s3/s3_rest.js:
@Neon-White Why are you changing this file `s3_rest` also? Are you sure?

guymguym commented on src/endpoint/s3/s3_utils.js:
```diff
@@ -622,6 +622,17 @@ function encode_uri_unless_already_encoded(uri = '') {
 
+/**
+ * parse_version_id throws an error if version_id is an empty string, and returns it otherwise
+ * @param {string} version_id
```
```suggestion
 * @param {string|undefined} version_id
```

guymguym commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -118,9 +119,7 @@ async function handle_request(req, res) {
-
-
-    if (req.query && req.query.versionId) {
+    if (s3_utils.parse_version_id(req.query?.versionId)) {
```
I don't think this should be changed. the entire point in this code block was to prevent versioning for caching bucket. I agree that this check as a whole should have been added differently, perhaps just in a separate function than inline like that in the rest api flow, but still, we do not need to affect it with validation of the version id.

guymguym commented on src/endpoint/s3/s3_rest.js:
```diff
@@ -266,7 +265,7 @@ function _get_method_from_req(req) {
         throw new S3Error(S3Error.InternalError);
     }
-    if (req.query && req.query.versionId && s3_op.versioned) {
+    if (s3_utils.parse_version_id(req.query?.versionId) && s3_op.versioned) {
```
again also here, the actual value of version id is not needed and should not be validated. the check was only if version id is set for the request. if you want you can add a separate `s3_utils.is_versioning_request(req)` which will just return `Boolean(req.query?.versionId)`

guymguym commented on src/endpoint/s3/ops/s3_post_bucket_delete.js:
```diff
@@ -29,7 +30,7 @@ async function post_bucket_delete(req) {
     for (const item of delete_list) {
         const key = item.Key?.[0];
-        const version_id = item.VersionId?.[0];
+        const version_id = s3_utils.parse_version_id(item.VersionId?.[0]);
```
did you check that AWS also errors in this api when sent with `<VersionId></VersionId>`?

guymguym replied:
I checked this one and this should be reverted, see other comment for details.

Neon-White committed:
- Improve `parse_version_id` docs
- Revert `s3_rest.js` changes

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
- Revert `s3_post_bucket_delete` changes
- Add empty version ID marker error
- Add empty version ID marker checks

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
Add empty version ID marker test

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
Fix test print

Signed-off-by: Ben <belimele@redhat.com>

shirady reviewed (approved):
LGTM

guymguym commented on src/endpoint/s3/s3_utils.js:
```diff
@@ -622,6 +622,28 @@ function encode_uri_unless_already_encoded(uri = '') {
     return is_uri_already_encoded(uri) ? uri : encodeURIComponent(uri);
 }
 
+/**
+ * parse_version_id throws an error if version_id is an empty string, and returns it otherwise
+ * @param {string|undefined} version_id
+ */
+function parse_version_id(version_id) {
+    if (version_id === '') {
+        throw new S3Error(S3Error.InvalidArgumentEmptyVersionId);
+    }
+    return version_id;
+}
+
+/**
+ * parse_version_id_marker throws an error if version_id_marker is an empty string, and returns it otherwise
+ * @param {string|undefined} version_id_marker
+ */
+function parse_version_id_marker(version_id_marker) {
+    if (version_id_marker === '') {
+        throw new S3Error(S3Error.InvalidArgumentEmptyVersionIdMarker);
+    }
+    return version_id_marker;
+}
+
```
I don't think we need two functions, this is handling the exact same type of information - a version_id. Since only the error is different you can have a default argument for all cases and for that different case of version_id_marker to pass the desired error spec from the caller as S3Error.InvalidArgumentEmptyVersionIdMarker

```suggestion
/**
 * parse_version_id throws an error if version_id is an empty string, and returns it otherwise
 * @param {string|undefined} version_id
 * @param {import('./s3_errors').S3ErrorSpec} [empty_err]
 */
function parse_version_id(version_id, empty_err = S3Error.InvalidArgumentEmptyVersionId) {
    if (version_id === '') {
        throw new S3Error(empty_err);
    }
    return version_id;
}

```

Neon-White committed:
Rework error flow

Signed-off-by: Ben <belimele@redhat.com>

Neon-White committed:
Mark `empty_err` as optional

Signed-off-by: Ben <belimele@redhat.com>

guymguym reviewed (approved):
Looks good!

nimrod-becker commented:
@guymguym 14 commits! not squashed :(

guymguym commented:
didnt notice. will be squashed by backports if needed.
